src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php

```diff
@@ -36,13 +38,56 @@
      *
      * @return Reference[]
      */
-    private function findAndSortTaggedServices($tagName, ContainerBuilder $container)
+    private function findAndSortTaggedServices($tagName, ContainerBuilder $container, string $indexAttribute = null, string $defaultIndexMethod = null)
     {
         $services = array();
 
+        if (null === $indexAttribute && null !== $defaultIndexMethod) {
+            throw new InvalidArgumentException('Default index method cannot be used without specifying a tag attribute.');
+        }
+
         foreach ($container->findTaggedServiceIds($tagName, true) as $serviceId => $attributes) {
             $priority = isset($attributes[0]['priority']) ? $attributes[0]['priority'] : 0;
-            $services[$priority][] = new Reference($serviceId);
+
+            if (null === $indexAttribute && null === $defaultIndexMethod) {
+                $services[$priority][] = new Reference($serviceId);
+
+                continue;
+            }
+
+            if (isset($attributes[0][$indexAttribute])) {
+                $services[$priority][$attributes[0][$indexAttribute]] = new Reference($serviceId);
+
+                continue;
+            }
+
+            if (null === $defaultIndexMethod) {
+                throw new LogicException(sprintf('Tag attribute with name "%s" for service "%s" cannot be found.', $indexAttribute, $serviceId));
+            }
+
+            if (!$r = $container->getReflectionClass($class = $container->getDefinition($serviceId)->getClass())) {
+                throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $serviceId));
+            }
+
+            if (!$r->hasMethod($defaultIndexMethod)) {
+                throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" has no method "%s".', $class, $serviceId, $defaultIndexMethod));
```

Comment:
here we could also hint that adding the attribute is another good option:
Method "%s::%s()" not found: Tag "%s" on service "%s" is missing a "%s" attribute. (or similar)